src/client/Client.cc

```diff
@@ -10740,6 +10745,76 @@ bool Client::is_quota_bytes_approaching(Inode *in)
   return false;
 }
 
+int Client::check_pool_caps(Inode *in, int need)
+{
+  int64_t pool = in->layout.fl_pg_pool;
+
+  int have = 0;
+  need &= CEPH_CAP_FILE_RD | CEPH_CAP_FILE_WR;
+
+  std::map<int64_t, int>::iterator it = pool_caps.find(pool);
+  if (it != pool_caps.end()) {
+    have = it->second;
+  } else {
+    char oid_buf[32];
+    snprintf(oid_buf, sizeof(oid_buf), "%llx.00000000", (unsigned long long)in->ino);
+    object_t oid = oid_buf;
+
+    Mutex lock("Client::check_pool_caps::lock");
+    Cond cond;
+    bool done[2] = {false, false};
+    int rd_ret;
+    int wr_ret;
+
+    ObjectOperation rd_op;
+    rd_op.stat(NULL, (utime_t*)NULL, NULL);
+
+    objecter->mutate(oid, OSDMap::file_to_object_locator(in->layout), rd_op,
+		     in->snaprealm->get_snap_context(), ceph_clock_now(cct), 0,
+		     new C_SafeCond(&lock, &cond, &done[0], &rd_ret), NULL);
+
+    ObjectOperation wr_op;
+    wr_op.create(true);
+
+    objecter->mutate(oid, OSDMap::file_to_object_locator(in->layout), wr_op,
+		     in->snaprealm->get_snap_context(), ceph_clock_now(cct), 0,
+		     new C_SafeCond(&lock, &cond, &done[1], &wr_ret), NULL);
+
+    client_lock.Unlock();
+    lock.Lock();
+    while (!done[0] || !done[1])
+      cond.Wait(lock);
+    lock.Unlock();
+    client_lock.Lock();
```

Comment:
This looks racy to me — certainly we could spin off somebody else checking if we can access the same pool, at minimum, but I'm concerned something more dangerous might be possible too. Like maybe getting caps revoked while we're in the process of testing? :/